File: 6.controllers-services-repositories-middlewares-and-request-context/code/typescript/a_note_on_nodejs_vs_go_python_1.ts
import { Request, Response } from 'express';

interface CreateBookRequest {
    title: string;
    author: string;
}

export function createBookHandler(req: Request, res: Response) {
    // Step 1 + 2: extract body and deserialize into a native class/dict.
    const payload = req.body as CreateBookRequest;
    if (!payload || !payload.title) {
        // Deserialization failed -> malformed payload.
        res.status(400).json({ error: "invalid request body" });
        return; // terminate the request here; do not proceed.
    }
    // ... validation, transformation, delegation follow ...
}

File: 6.controllers-services-repositories-middlewares-and-request-context/code/rust/a_note_on_nodejs_vs_go_python_1.rs
use actix_web::{web, HttpResponse, Responder};
use serde::Deserialize;

// CreateBookRequest is our native format ,  the bind target.
#[derive(Deserialize)]
pub struct CreateBookRequest {
    pub title: String,
    pub author: String,
}

pub async fn create_book_handler(req: web::Json<CreateBookRequest>) -> impl Responder {
    // Step 1 + 2: extract body and deserialize into struct handled by actix.
    // If it fails, actix automatically returns 400 Bad Request.
    let _payload = req.into_inner();
    
    // ... validation, transformation, delegation follow ...
    HttpResponse::Ok().finish()
}
